### DISCUSSION TITLE
How to create a GitHub Release in NukeBuild script / Octokit.Net

### DISCUSSION THREAD

**Discussion by Xilconic** [nuke-build/nuke#1358] [Q&A] [answered]
I want to create a [NukeBuild script](https://github.com/Xilconic/FlappyBirdUnityTutorial/blob/main/NukeBuild/Build.cs) that is capable of creating a GitHub Release for me (the `CreateReleaseOnGithub` Target in the linked build script). However I'm getting a `Oktokit.NotFoundException` without much else to go on then a URL to the documentation at https://docs.github.com/rest/releases/releases#create-a-release. Part of the stacktrace before it at my own custom build script calling `IReleaseClient.Create(string, string, NewReleaseData)`:
```
   at Octokit.Connection.HandleErrors(IResponse response) in /_/Octokit/Http/Connection.cs:line 780
   at Octokit.Connection.RunRequest(IRequest request, CancellationToken cancellationToken, Func`2 preprocessResponseBody) in /_/Octokit/Http/Connection.cs:line 761
   at Octokit.Connection.Run[T](IRequest request, CancellationToken cancellationToken, Func`2 preprocessResponseBody) in /_/Octokit/Http/Connection.cs:line 746
   at Octokit.ApiConnection.Post[T](Uri uri, Object data, String accepts, String contentType, CancellationToken cancellationToken)
```

For reference, consider this the code I'm executing within my Target:
```csharp
var productInformation = new ProductHeaderValue("MyApp-NukeBuildScript");
GitHubTasks.GitHubClient = new GitHubClient(productInformation)
{
    Credentials = new Credentials(GitHubUser, GitHubPassword) // My GitHub credentials I provide as command line arguments when executing the script
};

var releaseData = new NewRelease(Version)
{
    Prerelease = true,
	Draft = true,
    Body = "Test"
};

Release release = await GitHubTasks.GitHubClient.Repository.Release
	.Create(Repository.GetGitHubOwner(), Repository.GetGitHubName(), releaseData);
```

where `Repository` is a field defined within the class as
```csharp
[GitRepository] readonly GitRepository Repository;
```

I suspected I might have a typo in the `owner` or `name` arguments, and that this would result in not being able to resolve the repository I want to create the Release in. However in an experiment I've executed the following code in a NukeBuild target:
```csharp
string branchName = await GitHubTasks.GetDefaultBranch(Repository);
```
which is a piece of Code I pulled from the NukeBuild source code within the [GitHubTasks source](https://github.com/nuke-build/nuke/blob/011956b31c05f14f3233f6241cd6fbe038824d71/source/Nuke.Common/Tools/GitHub/GitHubTasks.cs#L61). And when I run this code, I successfully get the name of my repository.

So it seems that my problem is not related to the `Repository.GetGitHubOwner()` or `Repository.GetGitHubName()` arguments being passed to the `Create(string, string, ...)` method, which leaves me confused.

Why is my piece of code unable to create a Github Release?

**Comment by valadas** [comment]
If a working example helps, here is a public repo where I do it:

https://github.com/DNNCommunity/dnn-elements/blob/22b35de3b21e7975309d9968b7b5df5d65cfc33f/_build/Build.cs#L261

**Reply by Xilconic** [reply]
I noticed the overload of that you are referring to, but I didn't go for it because it was unclear to me what the `GithubToken` _is_.
I wasn't able to find any documentation about where I would be able to get it from or how to generate it.
@valadas : Would you be able to elaborate on that? Then I can give that overload a try instead and see what happens. :)

**Comment by valadas** [comment]
The github token is a token generate for a single run of a github action. I set it up like this to let github know to provide it to me https://github.com/DNNCommunity/dnn-elements/blob/22b35de3b21e7975309d9968b7b5df5d65cfc33f/_build/Build.cs#L34

Then we capture its value in a parameter like so: https://github.com/DNNCommunity/dnn-elements/blob/22b35de3b21e7975309d9968b7b5df5d65cfc33f/_build/Build.cs#L67-L68

Which allows us to do git operations that require authentication according to the permissions of that token generated for us by github.

In the repository settings for actions, you will need read/write permissions for that token to be able to publish releases.

![image](https://github.com/nuke-build/nuke/assets/6371568/3928e00b-7b78-445f-92f3-224f339096d1)

**Reply by Xilconic** [reply]
In my case I'm not making use of Github Actions. So what would be alternative for being able to do with, while running from a commandline on my own machine for example?

**Reply by valadas** [reply]
I would think roughly the same but using a Personal Access Token (PAT) https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens

Or potentially if you know the git user already has require permissions, then maybe no authentication at all. Just a guess though.

**Comment by Xilconic** [comment, marked as answer]
I have not been able to make it work with the 'username + password' overload of the `Credentials` class.

I did manage to get the intergration to work with a Github Personal Access Token.

 1. Create a Personal Access Token (PAT) in Github (source: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token) Make sure it has the grant for Reading and Writing of contents, in order to be authorized to create Github Releases (as well as uploading files to that release).
2. Make sure you store the PAT somewhere safe, like a password safe.
3. Define a Parameter in the NukeBuild script for passing in the PAT. Ensure you decorate the property with `[Secret]` as you are dealing with secret data here. From the root of the repository, run `nuke` on the commandline to build the script and populate the `build.schema.json` file in the `.nuke` folder.
4. In the `.nuke` folder, create a parameters file to store the data, for example `parameters.secrets.json`, with file contents `{}` (an empty json object). Consider putting this file in gitignore, depending on your situation.
5. From your repository root, open command line and run `nuke :secrets secrets`. Provide a password (also save this somewhere, as you'll need it to decrypt the file again later!).
6. Navigate the menu and select the PAT parameter you've defined. Fill in the PAT. Then `<Save and Exit>`. This now saves the PAT in encrypted format in that secrets file.
7. Run the buildscript target, providing the `secrets` profile: `nuke --target MyTarget --profile secrets`. You'll be asked the password for decrypting the secrets file (from step 5). Provide it, and the script will now create the Release in Github.
